src/lib/src/grid3d/grid.cpp

```diff
@@ -247,4 +247,66 @@ create_grid_from_cube(const cube::Cube &cube,
     return std::make_tuple(coordsv, zcornsv, actnumsv);
 }
 
+std::tuple<py::array_t<float>, py::array_t<int8_t>>
+refine_vertically(const Grid &grid_cpp, const py::array_t<int8_t> refinement_per_layer)
+{
+    auto actnumsv_ = grid_cpp.actnumsv.unchecked<3>();
+    auto zcornsv_ = grid_cpp.zcornsv.unchecked<4>();
+    auto refinement_data = refinement_per_layer.unchecked<1>();
+
+    // logging here, more for demonstration than anything else
+    auto &logger = xtgeo::logging::LoggerManager::get("xtgeo.grid3d.refine_vertically");
+
+    size_t total_refinement = 0;
+
+    for (size_t i = 0; i < refinement_data.shape(0); i++) {
+        total_refinement += refinement_data(i);
+    }
+
+    // Shape
+    std::vector<size_t> zcorn_shape = { grid_cpp.ncol + 1, grid_cpp.nrow + 1,
+                                        total_refinement + 1, 4 };
+    py::array_t<float> zcornref(zcorn_shape);
+    py::array_t<int8_t> actnumref({ grid_cpp.ncol, grid_cpp.nrow, total_refinement });
+    auto zcornref_ = zcornref.mutable_unchecked<4>();
+    auto actnumref_ = actnumref.mutable_unchecked<3>();
+
+    logger.debug("Refine vertically from {} layers to {} layers", grid_cpp.nlay,
+                 total_refinement);
+
+    for (size_t j = 0; j < grid_cpp.nrow; j++)
+        for (size_t i = 0; i < grid_cpp.ncol; i++) {
+            int kk = 0; /* refined grid K counter */
+            for (size_t k = 0; k < grid_cpp.nlay; k++) {
```

Comment:
I think you should use `j <= grid_cpp.nrow` (not `<` ) etc since zcorn has shape (ncol +1, nrow +1, nlay +1, 4).

Do an exception for actnum

Also I think `i` should form the outermost loop (C: k fastest, then j, then i)

Reply:
That solved the issue. Thanks